File: crdt/crdt-py/cloudy/gcounter.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

# A simple growth-only CRDT.
# See acobster's notes: https://acobster.keybase.pub/recurse/crdts

import logging

logger = logging.getLogger(__name__)
logging.basicConfig()
logger.setLevel(logging.INFO)

class GCounter(object):
    '''An increment-only CRDT store.'''

    def __init__(self, idx, pubsub):
        self.idx = idx
        self.ps = pubsub

        # State is a map of known node IDs to node values.
        self._state = {self.idx: 0}

        # Subscribe self for state updates
        self.ps.subscribe(self.idx, self._join)

    # Public interface

    def increment(self):
        '''Increment the counter for this instance.'''
        logger.info("Increment Node %d" % self.idx)
        self._state[self.idx] += 1
        self.ps.publish(state=self._state)

    def value(self):
        '''Request the current global "value", or sum of state counts'''
        return sum(v for k, v in self._state.items())

    # Internal state management

    def _join(self, state):
        '''Merge a published state update into our internal state.'''
        logger.debug("Node %d %s - received %s" % (self.idx, str(self._state), str(state)))
        self._state = {
            idx: max(state.get(idx, 0), self._state.get(idx, 0))
            for idx in set(state.keys()).union(self._state.keys())
        }
        logger.debug("\t->", self._state)

File: crdt/crdt-go/gcounter.go
package main

import "fmt"

// Interface for a node object that has a UID and can receive state updates
type Node interface {
	getID() int
	join(map[int]int) // callable to receive a state update
}

// Interface for publishing state updates to Node subscribers.
type PubSubInterface interface {
	publish(int, map[int]int)
}

// GCounter is a grow-only counter that satisfies the Node interface.
type GCounter struct {
	ID     int
	State  map[int]int
	PubSub PubSubInterface
}

// NodeManager communicates between Nodes and satisfies the PubSub interface.
type NodeManager struct {
	Nodes     map[int]Node
	maxIDSeen int // not guaranteed to currently be in Nodes
}

// Instantiate NodeManager state
func NewNodeManager() *NodeManager {
	return &NodeManager{maxIDSeen: -1, Nodes: make(map[int]Node)}
}

// Create and return a new GCounter node with autoincrementing ID
func (nm *NodeManager) newNode() *GCounter {
	// Create a GCounter with new unique ID and reference to NodeManager
	n := NewGCounter(nm.maxIDSeen+1, nm)
	// Increment our max ID seen to match
	nm.maxIDSeen++
	// Track in our list of Nodes
	nm.Nodes[n.ID] = n
	return n
}

// Sends out a state update to all tracked Nodes.
func (nm *NodeManager) publish(senderID int, state map[int]int) {
	for _, node := range nm.Nodes {
		if node.getID() != senderID { // We can skip publishing to the sender
			node.join(state)
		}
	}
}

// Instantiate GCounter state
func NewGCounter(id int, pubsub PubSubInterface) *GCounter {
	state := make(map[int]int)
	state[id] = 0
	return &GCounter{ID: id, State: state, PubSub: pubsub}
}

// getID() method to satisfy the Node interface.
func (g *GCounter) getID() int {
	return g.ID
}

// Increment the CRDT state via this GCounter node.
func (g *GCounter) increment() {
	// Increment this node's state
	g.State[g.ID] += 1
	// Publish to others
	fmt.Printf("\nNode %d increment()\n", g.ID)
	g.PubSub.publish(g.ID, g.State)
}

// Request the current global CRDT "value", or sum of known counts in state
func (g *GCounter) value() int {
	value := 0
	for _, count := range g.State {
		value += count
	}
	return value
}

// Merge a received state into our internal state.
func (g *GCounter) join(state map[int]int) {
	for id, count := range state {
		val, ok := g.State[id]
		// update count to the max of the received & existing count
		if ok {
			g.State[id] = max(count, val)
		} else {
			g.State[id] = count
		}
	}
	fmt.Printf("Node %d state: %v\n", g.ID, g.State)
}

// Return the maximum of two integers.
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// Sandbox
func main() {
	nm := NewNodeManager()

	node0 := nm.newNode()
	fmt.Println(node0.ID)
	fmt.Println(node0.State)
	node0.increment()
	fmt.Println(node0.State)

	node1 := nm.newNode()
	node1.increment()
	node1.increment()

	node0.increment()
}
